src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -31,6 +31,72 @@
 
 namespace Microsoft.PowerShell.Commands
 {
+    #region InodeTracker
+    /// <summary>
+    /// Tracks visited files/directories by caching their device IDs and inodes.
+    /// </summary>
+    internal class InodeTracker
+    {
+        private Dictionary<UInt64, Dictionary<UInt64, bool>> _visitations;
+
+        /// <summary>
+        /// Construct a new InodeTracker
+        /// </summary>
+        public InodeTracker()
+        {
+            _visitations = new Dictionary<UInt64, Dictionary<UInt64, bool>>();
+        }
+
+        /// <summary>
+        /// Determine if a path has been visited.
+        /// </summary>
+        /// <param name="path">Path to the file or directory to be checked.</param>
+        /// <returns>True if the path has been visited, false otherwise</returns>
+        public bool Visited(string path)
+        {
+            UInt64 dev;
+            UInt64 inode;
+
+            if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out dev, out inode))
+            {
+                if (_visitations.ContainsKey(dev))
+                {
+                    return _visitations[dev].ContainsKey(inode);
```

Comment:
It may be that on Windows the user will have only one volume, but that is much less the case on Unix. For example, on my Linux box the root (`/`) directory has device ID 2049, the `/dev` directory has ID 6, and the `/proc` directory has ID 4. Symlinks can cross device/volume boundaries.

I've changed the code to use the C#7 ValueTuple mechanism.